src/Components/Endpoints/src/Builder/RazorComponentsEndpointHttpContextExtensions.cs

```diff
@@ -0,0 +1,47 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Reflection;
+using System.Reflection.Metadata;
+using Microsoft.AspNetCore.Components.Endpoints;
+using Microsoft.AspNetCore.Components.Routing;
+using Microsoft.AspNetCore.Http;
+
+[assembly: MetadataUpdateHandler(typeof(RazorComponentsEndpointHttpContextExtensions.MetadataUpdateHandler))]
+
+namespace Microsoft.AspNetCore.Components.Routing;
+
+/// <summary>
+/// Extensions to <see cref="HttpContext"/> for Razor component applications.
+/// </summary>
+public static class RazorComponentsEndpointHttpContextExtensions
+{
+    private static readonly ConcurrentDictionary<Type, bool> AllowsInteractiveRoutingCache = new();
+
+    /// <summary>
+    /// Determines whether the current endpoint is a Razor component that can be reached through
+    /// interactive routing. This is true for all page components except if they declare the
+    /// attribute <see cref="AllowInteractiveRoutingAttribute"/> with value <see langword="false"/>.
+    /// </summary>
+    /// <param name="context">The <see cref="HttpContext"/>.</param>
+    /// <returns>True if the current endpoint is a Razor component that does not declare <see cref="AllowInteractiveRoutingAttribute"/> with value <see langword="false"/>.</returns>
+    public static bool AllowsInteractiveRouting(this HttpContext? context)
+    {
+        var pageType = context?.GetEndpoint()?.Metadata.GetMetadata<ComponentTypeMetadata>()?.Type;
```

Comment:
The attribute is going to be available directly in the endpoint metadata, no need to reach for the type below with GetCustomAttribute